File: docs/leetcode/39/39.py
# -*- coding:utf-8 -*-

from typing import List 

class Solution:
    def dfs(self, start, target, line, ans, candidates):
        if target == 0:
            ans.append(line[:])
            return

        for i in range(start, len(candidates)):
            if candidates[i] > target:
                break
            line.append(candidates[i])
            self.dfs(i, target - candidates[i], line, ans, candidates)
            line.pop()
        return

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        line = []
        ans = []
        if (len(candidates) == 0):
            return ans

        candidates = sorted(candidates)
        self.dfs(0, target, line, ans, candidates)

        return ans

s = Solution()
ans = s.combinationSum([2,7,6,3], 7)
print(ans)

File: docs/leetcode/39/39.cpp
#include <vector>
#include <string>
#include <stack>
#include <algorithm>
#include <deque>

using namespace std; 


class Solution {
public:
    void dfs(int start, int target, vector<int> &line, vector<vector<int>> &ans, vector<int>& candidates) {
        if (target == 0) {
            ans.push_back(line);
            return;
        }

        for (int i = start; i < candidates.size(); i++) {
            if (candidates[i] > target) {
                break;
            }
            line.push_back(candidates[i]);
            dfs(i, target - candidates[i], line, ans, candidates);
            line.pop_back();
        }
        return;
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        if (candidates.size() == 0) {
            return ans;
        }

        vector<int> line;
        sort(candidates.begin(), candidates.end());
        dfs(0, target, line, ans, candidates);

        return ans;

    }
};

int main()
{
    Solution *s = new Solution();
    vector<int> can = {2,3,6,7};
    int target = 7;
    auto ans = s->combinationSum(can, target);
    return 0;
}